### ISSUE TITLE
Improve IDE0046 to not report when in a else if chain

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn#59135]
**Brief description:**

I'd like to suggest to relax `IDE0046` (could be through some user option if we want to preserve the old behavior) to not report if the if to simplify is part of a if/else if chain.

**Languages applicable:**

Both.

**Code example that the analyzer should report:**

Currently, the rule reports on the following:

```csharp
if (t >= 0 && t <= 19)
{
	return 0x5A827999u;
}
else if (t >= 20 && t <= 39)
{
	return 0x6ED9EBA1u;
}
else if (t >= 40 && t <= 59)
{
	return 0x8F1BBCDCu;
}
else
{
	return t >= 60 && t <= 79 ? 0xCA62C1D6u : throw new ArgumentException("Argument out of bounds! 0 <= t < 80", nameof(t));
}
```

Applying the code-fix gives (IMO) a code less readable.

**Additional information:**

We could also relax the rule when the code-fix would result in a chained ternary. For example, applying all changes on the code above results in:

```csharp
return t >= 0 && t <= 19
                ? 0x5A827999u
                : t >= 20 && t <= 39
                    ? 0x6ED9EBA1u
                    : t >= 40 && t <= 59
                                    ? 0x8F1BBCDCu
                                    : t >= 60 && t <= 79 ? 0xCA62C1D6u : throw new ArgumentException("Argument out of bounds! 0 <= t < 80", nameof(t));
```

which again seems less readable to me.

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by CyrusNajmabadi** [comment]
Seems readable to me.  I'm not sure why if/else should be special cased out of this.

**Comment by Evangelink** [comment]
This exact case would be best with improved pattern matching but generally I don't think nested ternaries are easy to read/comprehend. 

Obviously, if I am alone thinking that, you can clause the ticket.

**Comment by CyrusNajmabadi** [comment]
Ultimately, i think it's entirely subjective.  Which means taht any time we make any changes we always have a risk of someone finding it less readable.  I think if that'st eh case, the best thing to do is just disable the feature.

**Comment by Evangelink** [comment]
I sure can but then I loose the part I like (i.e. refactor simple if/else with a ternary).

In the end, I understand that such change even if it seems trivial means that you will have to support it so I am fine if nothing happens.

**Comment by CyrusNajmabadi** [comment]
in that case, you can just ignore the issue in this location.  no need to apply teh suggestion in code where it's not valuable for you :)

**Comment by Evangelink** [comment]
True! I guess I was looking at a way to "enforce" some styles in the repo.

**Comment by jinujoseph** [comment]
In design review: To revist the original purpose of this and if the suggested change is diserable

**Comment by genlu** [comment]
Design meeting conclusion:

Will not change. This feature is intended to be used more like a refactoring but implemented as analyzer/codefix instead just to have some benefits of being an analyzer. If it's required to make this analyzer a build blocker, we'd suggest implementing a [DiagnosticSuppressor](https://sourceroslyn.io/Microsoft.CodeAnalysis/R/8afaec075a92603b.html) to fine tune the behavior as needed.

Additionally, https://github.com/dotnet/roslyn/issues/59352 is filed to track the improvement of refactoring UX to avoid getting into this situation again in the future.

**Comment by LeftofZen** [comment]
+1 from me. No point using this analyzer if it cannot handle basic cases like this and instead spits out worse code. Bizarre decision to not improve the analyzer too, imagine writing an analyzer, it being bad, then saying "nah write *more* code to suppress the shitty analyzer output.

**Comment by CyrusNajmabadi** [comment]
> +1 from me. No point using this analyzer if it cannot handle basic cases like this and instead spits out worse code. Bizarre decision to not improve the analyzer too. Imagine writing an analyzer, it being bad, then saying "nah write *more* code to suppress the shitty analyzer output".

There's no consensus this is "worse code". During the design, the feeling was the way that the was performing was the desirable way.  We don't make one off tweaks for smatterings of users. 

> then saying "nah write more code to suppress

That's literally how all analyzers work.  There are always a tiny handful of people that like things ever so slightly different. We still don't cater to such smart groups as it's not an effective use of time or resources.
